File: R/kwak_intensity.R
#**********************************************************************
#* 
#* intensity.R
#* 
#* Input
#*   xomat : xo matrix
#*   window : window size
#*   marker : marker vector
#*   n_ind : # of observations
#*   N : # of marker positions that we want to calculate
#*
#* Return : 
#*   val : intensity fcn value at each of marker positions
#*   center : marker position vector (It have length N)
#*   window_size : window size 
#*
#**********************************************************************/

intensity <-
function(cross, chr, window=2.5, ncalc=500)
{
  if(!missing(chr)) {
    cross <- subset(cross, chr)
    if(nchr(cross) > 1)
      warning("Considering only chr ", names(cross$geno)[1])
  }

  if(class(cross)[1] != "bc")
    stop("coincidence() currently working only for a backcross.")

  g <- cross$geno[[1]]$data
  g[is.na(g)] <- 0
  map <- cross$geno[[1]]$map

  xomat <- identify_xo(g)$xomat
  n.ind <- nind(cross)

  intensitySUB(xomat, window, map, n.ind, ncalc)
}


intensitySUB <- 
function(xomat, window, marker, n_ind, N)
{

  n_xo <- ncol(xomat)
  n_pos <- length(marker)

  center <- seq(marker[1], marker[n_pos], length=N)

  n_center <- length(center)
  xovec <- c(xomat)

  output <- .C("R_get_intensity",
               as.integer(xovec),
               as.double(window),
               as.double(center),
               as.integer(n_pos),
               as.integer(n_xo),
               as.integer(n_center),
               as.double(marker),
               intensity=as.double(rep(0,n_center)),
               as.integer(n_ind),
               PACKAGE="xoi")

  result <- data.frame(position=center,
                       intensity=output$intensity)
  attr(result, "window") <- window
  result
}

# end of intensity.R

File: src/kwak_intensity.c
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include <R_ext/Applic.h>
#include <R_ext/Utils.h>
#include "kwak.h"


/**********************************************************************
 *
 * get_intensity
 *
 * get intensity probability for each divided marker interval.
 *
 **********************************************************************/
void R_get_intensity(int *xovec, double *window, double *center,
                     int *n_pos, int *n_xo,  int *n_center,
                     double *marker, double *intensity, int *n_ind)
{
    get_intensity(xovec, *window, center, *n_pos, *n_xo, *n_center, marker, intensity, *n_ind);
}

void get_intensity(int *xovec, double window, double *center,
                   int n_pos, int n_xo, int n_center, double *marker,
                   double *intensity, int n_ind)
{
    int i, j, ob;
    double weight, prob;

    for(i = 0 ; i < n_center ; i++)
        {
            ob = xovec[0];
            prob = 0;
            for(j = 0 ; j < n_xo ; j++)
                {
                    if(ob == xovec[j*3])
                        {
                            if( ( marker[xovec[j*3+1]-1] <= center[i] + window/2 &&
                                  marker[xovec[j*3+1]-1] >= center[i] - window/2 ) ||
                                ( marker[xovec[j*3+2]-1] <= center[i] + window/2 &&
                                  marker[xovec[j*3+2]-1] >= center[i] - window/2 ) ||
                                ( marker[xovec[j*3+1]-1] <= center[i] - window/2 &&
                                  marker[xovec[j*3+2]-1] >= center[i] + window/2 )  )
                                {
                                    weight = MIN(center[i]+window/2, marker[xovec[j*3+2]-1]) - MAX(center[i]-window/2, marker[xovec[j*3+1]-1]) ;
                                    prob += weight / ( marker[xovec[j*3+2]-1] - marker[xovec[j*3+1]-1] );
                                    //    Rprintf("j : %d, (%f,%f) ob : %d,  weight : %f,  prob : %f \n", j, center[i] - window/2, center[i] + window/2, ob, weight, prob);
                                }
                        } else {
                        if (prob > 1) prob = 1;
                        intensity[i] = intensity[i] +  prob;
                        //    Rprintf("(%f,%f) ob : %d,  intensity[%i] : %f \n", center[i] - window/2, center[i] + window/2, ob, i, intensity[i]);
                        ob = xovec[j*3];
                        prob = 0;
                        if( ( marker[xovec[j*3+1]-1] <= center[i] + window/2 &&
                              marker[xovec[j*3+1]-1] >= center[i] - window/2 ) ||
                            ( marker[xovec[j*3+2]-1] <= center[i] + window/2 &&
                              marker[xovec[j*3+2]-1] >= center[i] - window/2 ) ||
                            ( marker[xovec[j*3+1]-1] <= center[i] - window/2 &&
                              marker[xovec[j*3+2]-1] >= center[i] + window/2 )        )
                            {
                                weight = MIN(center[i]+window/2, marker[xovec[j*3+2]-1]) - MAX(center[i]-window/2, marker[xovec[j*3+1]-1]) ;
                                prob += weight / ( marker[xovec[j*3+2]-1] - marker[xovec[j*3+1]-1] );
                                //    Rprintf("(%f,%f) ob : %d,  weight : %f,  prob : %f \n", center[i] - window/2, center[i] + window/2, ob, weight, prob);
                                if (j==(n_xo-1))
                                    {
                                        if (prob > 1) prob = 1;
                                        intensity[i] = intensity[i] +  prob;
                                        //        Rprintf("j : %d (%f,%f) ob : %d,  intensity[%i] : %f \n", j, center[i] - window/2, center[i] + window/2, ob, i, intensity[i]);
                                    }
                            }
                    }
                }

            /* contained window */
            weight = MIN(center[i]+window/2, marker[n_pos-1]) - MAX(center[i]-window/2, marker[0]);

            intensity[i] /= (weight*(double)n_ind/100.0);

        } /* loop over center */
}
